src/main/java/codechicken/nei/ItemsGrid.java

```diff
@@ -244,6 +257,121 @@ private void drawSlotOutlines(int mousex, int mousey) {
         }
     }
 
+    private int recipeTooltipSlotIdx = -1;
+    private int recipeTooltipLines = 2;
+    private Runnable recipeTooltipUpdater = null;
+    private GuiRecipe<?> recipeTooltipGui = null;
+
+    public void update() {
+        if (recipeTooltipUpdater != null) {
+            recipeTooltipUpdater.run();
+            recipeTooltipUpdater = null;
+        }
+        if (recipeTooltipGui != null) {
+            recipeTooltipGui.updateAsTooltip();
+        }
+    }
+
+    private static <T> List<T> listOrEmptyList(final List<T> listOrNull) {
+        return listOrNull == null ? Collections.emptyList() : listOrNull;
+    }
+
+    private void drawRecipeTooltip(int mousex, int mousey) {
+        final Minecraft mc = Minecraft.getMinecraft();
+        ItemPanelSlot focused = getSlotMouseOver(mousex, mousey);
+        if (focused == null) {
+            recipeTooltipSlotIdx = -1;
+            recipeTooltipGui = null;
+            return;
+        }
+        final List<Integer> mask = getMask();
+
+        int slotIdx = -1;
+        for (int i = 0; i < mask.size(); i++) {
+            if (mask.get(i) == null) {
+                continue;
+            }
+            if (focused.slotIndex != mask.get(i)) {
+                continue;
+            }
+            slotIdx = i;
+            break;
+        }
+        if (slotIdx == -1) {
+            recipeTooltipSlotIdx = -1;
+            recipeTooltipGui = null;
+            return;
+        }
+
+        final Point mouseover = getMousePosition();
+        final ItemPanelSlot panelSlot = ItemPanels.bookmarkPanel.getSlotMouseOver(mouseover.x, mouseover.y);
+        final BookmarkRecipeId recipeId;
+        if (panelSlot != null) {
+            recipeId = ItemPanels.bookmarkPanel.getBookmarkRecipeId(panelSlot.slotIndex);
+        } else {
+            recipeId = ItemPanels.bookmarkPanel.getBookmarkRecipeId(focused.item);
+        }
+        if (recipeId == null) {
+            return;
+        }
+
+        if (slotIdx != recipeTooltipSlotIdx) {
+            recipeTooltipSlotIdx = slotIdx;
+            recipeTooltipGui = null;
+            recipeTooltipUpdater = () -> {
+                recipeTooltipGui = GuiCraftingRecipe.createRecipeGui("item", false, false, false, focused.item);
+                if (recipeTooltipGui != null) {
+                    recipeTooltipGui.limitToOneRecipe();
+                    recipeTooltipGui.initGui(true);
+                    recipeTooltipGui.guiTop = 0;
+                    recipeTooltipGui.guiLeft = 0;
+                }
+                recipeTooltipLines = 2;
```

Comment:
Why start with 2 here?

Reply:
Item name + Mod name lines, this only really matters if for some reason the tooltip text getter breaks completely

Reply:
The mod name in tooltips is not from vanilla forge, but added by WAILA, which is not a dependency of NEI, we should assume 1 as a starting point. It doesn't sound like there is a fundamental necessity to start with 2.

(I hope I'm not annoying, I'm just trying to keep in mind that NEI is supposed to be used outside of GTNH).

Reply:
Sorry, this may be the wrong place, here, but I was asking because in my video you can see that the recipe widget assumes the tooltips are two lines when they are just one. Maybe this is an issue with (`GuiContainerManager#itemDisplayNameMultiline`)?